File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public bool IsBalanced(TreeNode root) {
        return CheckHeight(root) != -1;
    }

    private int CheckHeight(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int leftHeight = CheckHeight(root.left);
        if (leftHeight == -1) return -1;

        int rightHeight = CheckHeight(root.right);
        if (rightHeight == -1) return -1;

        if (Math.Abs(leftHeight - rightHeight) > 1) {
            return -1;
        }

        return Math.Max(leftHeight, rightHeight) + 1;
    }
}

File: Data Structures & Algorithms/depth-of-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        self.max_depth = 0
        self.calculate_max_depth(root, 1)

        return self.max_depth

    def calculate_max_depth(self, root: Optional[TreeNode], depth: int):
        if not root:
            return None
        
        self.max_depth = max(depth, self.max_depth)
        self.calculate_max_depth(root.left, depth + 1)
        self.calculate_max_depth(root.right, depth + 1)
        